src/hotspot/share/jfr/periodic/sampling/jfrCPUTimeThreadSampler.cpp

```diff
@@ -0,0 +1,760 @@
+  // the sampling period might be too low for the current Linux configuration
+  // so samples might be skipped and we have to compute the actual period
+  int64_t period = get_sampling_period() * (info->si_overrun + 1);
+  request._cpu_time_period = Ticks(period / 1000000000.0 * JfrTime::frequency()) - Ticks(0);
+  sample_thread(request._request, context, jt, tl);
+
+  if (queue.enqueue(request)) {
+    if (queue.size() == 1) {
+      tl->set_has_cpu_time_jfr_requests(true);
+      SafepointMechanism::arm_local_poll_release(jt);
+    }
+  } else {
+    queue.increment_lost_samples();
+  }
+
+  if (jt->thread_state() == _thread_in_native) {
+      if (!tl->wants_async_processing_of_cpu_time_jfr_requests()) {
+        tl->set_do_async_processing_of_cpu_time_jfr_requests(true);
+        JfrCPUTimeThreadSampling::trigger_async_processing_of_cpu_time_jfr_requests();
+      }
+  } else {
+    tl->set_do_async_processing_of_cpu_time_jfr_requests(false);
+  }
+
+  tl->release_cpu_time_jfr_queue_lock();
+}
+
+static const int SIG = SIGPROF;
+
+static void set_timer_time(timer_t timerid, int64_t period_nanos) {
+  struct itimerspec its;
+  if (period_nanos == 0) {
+    its.it_interval.tv_sec = 0;
+    its.it_interval.tv_nsec = 0;
+  } else {
+    its.it_interval.tv_sec = period_nanos / NANOSECS_PER_SEC;
+    its.it_interval.tv_nsec = period_nanos % NANOSECS_PER_SEC;
+  }
+  its.it_value = its.it_interval;
+  if (timer_settime(timerid, 0, &its, nullptr) == -1) {
+    warning("Failed to set timer for thread sampling: %s", os::strerror(os::get_last_error()));
+  }
+}
+
+bool JfrCPUSamplerThread::create_timer_for_thread(JavaThread* thread, timer_t& timerid) {
+  timer_t t;
+  struct sigevent sev;
+  sev.sigev_notify = SIGEV_THREAD_ID;
+  sev.sigev_signo = SIG;
+  sev.sigev_value.sival_ptr = &t;
+  ((int*)&sev.sigev_notify)[1] = thread->osthread()->thread_id();
+  clockid_t clock;
+  int err = pthread_getcpuclockid(thread->osthread()->pthread_id(), &clock);
+  if (err != 0) {
+    log_error(jfr)("Failed to get clock for thread sampling: %s", os::strerror(err));
+    return false;
+  }
+  if (timer_create(clock, &sev, &t) < 0) {
+    return false;
+  }
+  int64_t period = get_sampling_period();
+  if (period != 0) {
+    set_timer_time(t, period);
+  }
+  timerid = t;
+  return true;
+}
+
+
+void JfrCPUSamplerThread::stop_signal_handlers() {
+  // set the stop signal bit
+  Atomic::or_then_fetch(&_active_signal_handlers, STOP_SIGNAL_BIT, memory_order_acq_rel);
+  while (Atomic::load_acquire(&_active_signal_handlers) > STOP_SIGNAL_BIT) {
+    // wait for all signal handlers to finish
+    os::naked_short_nanosleep(1000);
+  }
+}
+
+// returns false if the stop signal bit was set, true otherwise
+bool JfrCPUSamplerThread::increment_signal_handler_count() {
+  // increment the count of active signal handlers
+  u4 old_value = Atomic::fetch_then_add(&_active_signal_handlers, (u4)1, memory_order_acq_rel);
+  if ((old_value & STOP_SIGNAL_BIT) != 0) {
```

Comment:
Combining stop signal with a counter is nice, you can then use `Atomic::cmpxchg` to avoid incrementing counter when the stop bit is set.

Reply:
I don't see how `Atomic::cmpxchg` would make the code easier.

Reply:
With my current code, I avoid having a loop, and in the fast path, I only have one atomic instruction.